numpy/lib/arraypad.py

```diff
@@ -69,821 +81,475 @@ def _round_ifneeded(arr, dtype):
-        if not `arr.dtype` will be cast to `arr.dtype`.
+    shape : tuple
+        Shape of the array to slice as returned by its `shape` attribute.
+    sl : slice
+        The slice for the given dimension.
     axis : int
-        Axis along which to pad `arr`.
+        The axis to which `sl` is applied. All other dimensions are left
+        "unsliced".
 
     Returns
     -------
-    padarr : ndarray
-        Output array, with `pad_amt` constant `val` prepended along `axis`.
-
-    """
-    if pad_amt == 0:
-        return arr
-    padshape = tuple(x if i != axis else pad_amt
-                     for (i, x) in enumerate(arr.shape))
-    return _do_prepend(arr, np.full(padshape, val, dtype=arr.dtype), axis)
+    sl : tuple of slices
+        A tuple with slices matching `shape` in length.
 
-
-def _append_const(arr, pad_amt, val, axis=-1):
-    """
-    Append constant `val` along `axis` of `arr`.
-
-    Parameters
-    ----------
-    arr : ndarray
-        Input array of arbitrary shape.
-    pad_amt : int
-        Amount of padding to append.
-    val : scalar
-        Constant value to use. For best results should be of type `arr.dtype`;
-        if not `arr.dtype` will be cast to `arr.dtype`.
-    axis : int
-        Axis along which to pad `arr`.
-
-    Returns
-    -------
-    padarr : ndarray
-        Output array, with `pad_amt` constant `val` appended along `axis`.
-
-    """
-    if pad_amt == 0:
-        return arr
-    padshape = tuple(x if i != axis else pad_amt
-                     for (i, x) in enumerate(arr.shape))
-    return _do_append(arr, np.full(padshape, val, dtype=arr.dtype), axis)
-
-
-
-def _prepend_edge(arr, pad_amt, axis=-1):
-    """
-    Prepend `pad_amt` to `arr` along `axis` by extending edge values.
-
-    Parameters
-    ----------
-    arr : ndarray
-        Input array of arbitrary shape.
-    pad_amt : int
-        Amount of padding to prepend.
-    axis : int
-        Axis along which to pad `arr`.
-
-    Returns
-    -------
-    padarr : ndarray
-        Output array, extended by `pad_amt` edge values appended along `axis`.
-
-    """
-    if pad_amt == 0:
-        return arr
-
-    edge_slice = _slice_first(arr.shape, 1, axis=axis)
-    edge_arr = arr[edge_slice]
-    return _do_prepend(arr, edge_arr.repeat(pad_amt, axis=axis), axis)
-
-
-def _append_edge(arr, pad_amt, axis=-1):
+    Examples
+    --------
+    >>> _slice_at_axis((3, 4, 5), slice(None, 3, -1), 1)
+    (slice(None, None, None), slice(None, 3, -1), slice(None, None, None))
     """
-    Append `pad_amt` to `arr` along `axis` by extending edge values.
-
-    Parameters
-    ----------
-    arr : ndarray
-        Input array of arbitrary shape.
-    pad_amt : int
-        Amount of padding to append.
-    axis : int
-        Axis along which to pad `arr`.
+    slice_tup = (slice(None),)
+    return slice_tup * axis + (sl,) + slice_tup * (len(shape) - axis - 1)
```

Comment:
The last term `slice_tup * (len(shape) - axis - 1)` is optional. The tests still pass if removed. I think the explanation can be found in [NumPy's indexing guide](https://www.numpy.org/devdocs/user/basics.indexing.html#single-element-indexing):

> That is, each index specified selects the array corresponding to the rest of the dimensions selected. In the above example, choosing 0 means that the remaining dimension of length 5 is being left unspecified, and that what is returned is an array of that dimensionality and size.

Removing this term allows us to simplify the code even further by removing the now obsolete parameter `shape`. Because I can't think of a scenario where the last term might be needed I'm in favor. Do you agree?

Reply:
Yes, it is not needed. In a previous commit it was decided to keep it, so that the indexing is complete (in principle you could manipulate the index more). Both is good. I slightly lean to leaving it for now, unless it simplifies more then only these calls.

Reply:
Only the calls to `_slice_at_axis` and the function itself. 

You're talking about https://github.com/numpy/numpy/pull/11012#discussion_r189805848? A compromise might be to use `(...,)` as a placeholder for the last term. The indexing argument should then be ["complete"](https://docs.python.org/3/library/stdtypes.html#the-ellipsis-object) and the parameter `shape` would still be obsolete. I don't see any disadvantages to this approach.

Reply:
Yeah, probably was that comment. And fair point, if anyone attempts to mess with the tuple it will gracefully error out and can easily be changed back.

Reply:
Note: Python 2.7 doesn't support the syntax `...` outside sequence slicings. However as this PR won't make it into the 1.16 release (last one to support 2.7) this should be fine.

Reply:
Note that `np.s_[...,]` exists to solve that problem

Reply:
Note that `np.s_[...,]` exists to solve that problem

Reply:
Note that `np.s_[...,]` exists to solve that problem